edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/SubscriptionProcessor.cs

```diff
@@ -0,0 +1,256 @@
+            return Task.CompletedTask;
+        }
+
+        public Task ProcessSubscriptions(string id, IEnumerable<(DeviceSubscription, bool)> subscriptions)
+        {
+            Preconditions.CheckNonWhiteSpace(id, nameof(id));
+            List<(DeviceSubscription, bool)> subscriptionsList = Preconditions.CheckNotNull(subscriptions, nameof(subscriptions)).ToList();
+            Events.ProcessingSubscriptions(id, subscriptionsList);
+            foreach ((DeviceSubscription deviceSubscription, bool addSubscription) in subscriptionsList)
+            {
+                if (addSubscription)
+                {
+                    this.connectionManager.AddSubscription(id, deviceSubscription);
+                }
+                else
+                {
+                    this.connectionManager.RemoveSubscription(id, deviceSubscription);
+                }
+            }
+
+            this.AddToPendingSubscriptions(id, subscriptionsList);
+            return Task.CompletedTask;
+        }
+
+        internal async Task ProcessSubscription(string id, Option<ICloudProxy> cloudProxy, DeviceSubscription deviceSubscription, bool addSubscription)
+        {
+            Events.ProcessingSubscription(id, deviceSubscription);
+            try
+            {
+                switch (deviceSubscription)
+                {
+                    case DeviceSubscription.C2D:
+                        if (addSubscription)
+                        {
+                            cloudProxy.ForEach(c => c.StartListening());
+                        }
+
+                        break;
+
+                    case DeviceSubscription.DesiredPropertyUpdates:
+                        await cloudProxy.ForEachAsync(c => addSubscription ? c.SetupDesiredPropertyUpdatesAsync() : c.RemoveDesiredPropertyUpdatesAsync());
+                        break;
+
+                    case DeviceSubscription.Methods:
+                        if (addSubscription)
+                        {
+                            await cloudProxy.ForEachAsync(c => c.SetupCallMethodAsync());
+                            await this.invokeMethodHandler.ProcessInvokeMethodSubscription(id);
+                        }
+                        else
+                        {
+                            await cloudProxy.ForEachAsync(c => c.RemoveCallMethodAsync());
+                        }
+
+                        break;
+
+                    case DeviceSubscription.ModuleMessages:
+                    case DeviceSubscription.TwinResponse:
+                    case DeviceSubscription.Unknown:
+                        // No Action required
+                        break;
+                }
+            }
+            catch (Exception ex)
+            {
+                Events.ErrorProcessingSubscription(ex, id, deviceSubscription, addSubscription);
+            }
+        }
+
+        async void DeviceConnected(object sender, EventArgs eventArgs)
```

Comment:
This fires when the Edge Hub itself is connected to IoT Hub? We then process all existing subscriptions again?

Reply:
DeviceConnected is fired when connection to IoTHub is detected (for _any_ client). 
So when EdgeHub starts and and EdgeHub connects to IoTHub, this will be fired. However at that time, no other clients are connected, so we have no subscriptions to process.